org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/rename/LSPRenameHandler.java

```diff
@@ -97,19 +95,21 @@ public void setEnabled(Object evaluationContext) {
 			IDocument document = LSPEclipseUtils.getDocument((ITextEditor) part);
 			if (document != null && isEnable) {
 				try {
-					isEnable = !LanguageServiceAccessor.getLanguageServers(document, LSPRenameHandler::isRenameProvider)
-							.get(50, TimeUnit.MILLISECONDS).isEmpty();
+					// See if there is a server that supports rename. We don't use LanguageServers.anyMatching in this
+					// case because we want to time out and not block the UI if we have to wait for servers to start
+					// but we also want to be able to detect the timeout and schedule a handler update for when
+					// the definitive info on server capabilities is available
+					isEnable = LanguageServers.forDocument(document).withFilter(LSPRenameHandler::isRenameProvider)
+							.computeFirst(ls -> CompletableFuture.completedFuture(true))
+							.get(50, TimeUnit.MILLISECONDS).isPresent();
 				} catch (java.util.concurrent.ExecutionException | TimeoutException e) {
-
-					// in case the language servers take longer to kick in, defer the enablement to
-					// a later time
-					LanguageServiceAccessor.getLanguageServers(document, LSPRenameHandler::isRenameProvider)
-							.thenAccept(languageServer -> {
-								boolean enabled = !languageServer.isEmpty();
-								final var handleEvent = new HandlerEvent(this, enabled, false);
-								fireHandlerChanged(handleEvent);
-							});
-
+					// Timed out, schedule a handler update from when all servers are started and have provided their capabilities
```

Comment:
No your mistake, but should we not handle only `TimeoutException` with a retry, and not the `ExecutionException`?

Reply:
I think you're right - if there's an execution exception thrown here then it'll be thrown on retry (probably) anyway. Pulled that out into a pure error clause.